src/Symfony/Component/Messenger/MessageBus.php

```diff
@@ -41,37 +52,43 @@ public function __construct(iterable $middlewareHandlers = array())
     public function dispatch($message)
     {
         if (!\is_object($message)) {
-            throw new InvalidArgumentException(sprintf('Invalid type for message argument. Expected object, but got "%s".', \gettype($message)));
+            throw new \TypeError(sprintf('Invalid argument provided to "%s()": expected object, but got %s.', __METHOD__, \gettype($message)));
         }
 
-        return \call_user_func($this->callableForNextMiddleware(0, Envelope::wrap($message)), $message);
-    }
-
-    private function callableForNextMiddleware(int $index, Envelope $currentEnvelope): callable
-    {
-        if (null === $this->indexedMiddlewareHandlers) {
-            $this->indexedMiddlewareHandlers = \is_array($this->middlewareHandlers) ? array_values($this->middlewareHandlers) : iterator_to_array($this->middlewareHandlers, false);
+        $middlewareIterator = $this->middlewareAggregate->getIterator();
+        while ($middlewareIterator instanceof \IteratorAggregate) {
```

Comment:
Why `while`? What's the use-case in which it will return an iterator twice (or more)?

Reply:
@sroze i suggested that, see https://github.com/symfony/symfony/pull/28907#discussion_r226035693

Reply:
but maybe we should enforce it resolves ones, either here or in the constructor :thinking:

Reply:
That's how "foreach" works internally. We must havr it to preserve basic PHP semantics of Traversable.